thrust/cmake/thrust-config.cmake

```diff
@@ -0,0 +1,322 @@
+endif()
+
+if (THRUST_IGNORE_CUB_VERSION_CHECK)
+  target_compile_definitions(_Thrust_Thrust_Impl INTERFACE "THRUST_IGNORE_CUB_VERSION_CHECK")
+endif()
+
+#
+# Find and define backend target aliases:
+#
+
+# Update component state based on host/device requests:
+if (NOT "NONE" STREQUAL "${THRUST_HOST_SYSTEM}")
+  list(APPEND ${CMAKE_FIND_PACKAGE_NAME}_FIND_COMPONENTS ${THRUST_HOST_SYSTEM})
+  set(${CMAKE_FIND_PACKAGE_NAME}_FIND_REQUIRED_${THRUST_HOST_SYSTEM} ON)
+endif()
+if (NOT "NONE" STREQUAL "${THRUST_DEVICE_SYSTEM}")
+  list(APPEND ${CMAKE_FIND_PACKAGE_NAME}_FIND_COMPONENTS ${THRUST_DEVICE_SYSTEM})
+  set(${CMAKE_FIND_PACKAGE_NAME}_FIND_REQUIRED_${THRUST_DEVICE_SYSTEM} ON)
+endif()
+list(REMOVE_DUPLICATES ${CMAKE_FIND_PACKAGE_NAME}_FIND_COMPONENTS)
+
+# Prep helper variables used to configure backends:
+foreach(component ${${CMAKE_FIND_PACKAGE_NAME}_FIND_COMPONENTS})
+  # component is requested:
+  set(_THRUST_REQUESTS_${component} ON)
+  if (${CMAKE_FIND_PACKAGE_NAME}_FIND_REQUIRED_${component})
+    # component is required:
+    set(_THRUST_REQUIRED_FLAG_${component} "REQUIRED")
+  endif()
+endforeach()
+
+# Create Thrust::${backend}::Host and Thrust::${backend}::Device targets.
+# Checks if ${backend} matches the requested host/device backend, and if so,
+# links the corresponding host/device target to Thrust::Thrust.
+function(_thrust_setup_host_device_targets backend)
+  set(backend_target "Thrust::${backend}")
+
+  if (backend IN_LIST THRUST_HOST_SYSTEM_OPTIONS)
+    set(host_target "_Thrust_${backend}_Host")
+    set(host_target_alias "Thrust::${backend}::Host")
+    _thrust_declare_interface_alias(${host_target_alias} ${host_target})
+    target_compile_definitions(${host_target} INTERFACE
+      "THRUST_HOST_SYSTEM=THRUST_HOST_SYSTEM_${backend}")
+    target_link_libraries(${host_target} INTERFACE Thrust::Thrust::Impl)
+    target_link_libraries(${host_target} INTERFACE ${backend_target})
+
+    if ("${backend}" STREQUAL "${THRUST_HOST_SYSTEM}")
+      target_link_libraries(_Thrust_Thrust INTERFACE ${host_target_alias})
+    endif()
+  endif()
+
+  if (backend IN_LIST THRUST_DEVICE_SYSTEM_OPTIONS)
+    set(device_target "_Thrust_${backend}_Device")
+    set(device_target_alias "Thrust::${backend}::Device")
+    _thrust_declare_interface_alias(${device_target_alias} ${device_target})
+    target_compile_definitions(${device_target} INTERFACE
+      "THRUST_DEVICE_SYSTEM=THRUST_DEVICE_SYSTEM_${backend}")
+    target_link_libraries(${device_target} INTERFACE Thrust::Thrust::Impl)
+    target_link_libraries(${device_target} INTERFACE ${backend_target})
+
+    if ("${backend}" STREQUAL "${THRUST_DEVICE_SYSTEM}")
+      target_link_libraries(_Thrust_Thrust INTERFACE ${device_target_alias})
+    endif()
+  endif()
+endfunction()
+
+# Always provide CPP target:
+_thrust_declare_interface_alias(Thrust::CPP _Thrust_CPP)
+_thrust_setup_host_device_targets(CPP)
+
+# Setup CUDA targets
```

Comment:
What's the failure mode when someone requests the CUDA backend and they haven't done `enable_language(CUDA)`?

Reply:
Configure error:

```
find_package(Thrust REQUIRED)
thrust_create_target(Thrust DEVICE CUDA)
add_executable(main main.cu)
```

```
>cmake .
-- C++ Standard version: 14
-- Configuring done
CMake Error: CMake can not determine linker language for target: main
-- Generating done
CMake Generate step failed.  Build files cannot be regenerated correctly.
```

I think that's reasonable enough, it's the same thing that happens anything a `cu` file is compiled without enabling CUDA, so the error is very googleable.

I also checked to see if I could detect whether `enable_language(CUDA)` had been called, but there is no API for that :/ So we can't issue a conditional warning.